asset.go

```diff
@@ -0,0 +1,173 @@
+package main
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"strings"
+)
+
+func copyRegularFile(src, dest string) error {
+	srcFile, err := os.Open(src)
+	if err != nil {
+		return err
+	}
+	destFile, err := os.Create(dest)
+	if err != nil {
+		return err
+	}
+	if _, err := io.Copy(destFile, srcFile); err != nil {
+		return err
+	}
+	return nil
+}
+
+// makeAbsoluteTarget takes a absolute path which will serve as a base
+// and a relative symlink target to create absolute equivalent of
+// relative target.
+func makeAbsoluteTarget(base, target string) string {
+	dirPart := filepath.Dir(base)
+	testPath := filepath.Join(dirPart, target)
+	return filepath.Clean(testPath)
+}
+
+// absoluteSrcTargetToDest returns an absolute symlink target pointing
+// to an asset on ACI image, which is a counterpart of an absolute
+// symlink target pointing to an asset on local filesystem.
+//
+// Example:
+//
+// Given an asset /assets:$HOME/some/assets and a symlink somewhere
+// inside local asset which points to another place in local asset
+// like:
+// $HOME/some/assets/some/symlink -> $HOME/some/assets/some/target
+// This function returns target /assets/some/target. In this case
+// srcBase is $HOME/some/assets, srcTarget is
+// $HOME/some/assets/some/target and destBase is /assets.
+func absoluteSrcTargetToDest(srcBase, srcTarget, destBase string) (string, error) {
+	relTarget, err := filepath.Rel(srcBase, srcTarget)
+	if err != nil {
+		return "", err
+	}
+	destTarget := filepath.Join(destBase, relTarget)
+	return filepath.Clean(destTarget), nil
+}
+
+// copySymlink copies the symlink, but before doing so it ensures that
+// the symlink is pointing to node inside the asset.
+//
```

Comment:
Why is it necessary to check whether the symlink point outside the asset? You are creating the symlinks without dereferencing them, so there is no risks to copy something which does not belong to the asset in the ACI.

It prevents the user from creating a symlink to the compiled binary (which is not an asset) or to a file in a different asset (when several assets are given in parameters). I'm thinking of projects like Busybox with one binary and several symlinks to the same binary. The symlink ``$HOME/some/asset/mount`` pointing to ``/busybox`` would be broken on the build machine (let's say busybox is not installed on the build machine, or at least not at this path...) but once the ACI is built, it would not be broken in the container (because ``/busybox`` would be provided by the compilation).